src/ripple/core/impl/Workers.cpp

```diff
@@ -144,27 +143,50 @@ void Workers::deleteWorkers (beast::LockFreeStack <Worker>& stack)
 //------------------------------------------------------------------------------
 
 Workers::Worker::Worker (Workers& workers, std::string const& threadName)
-    : Thread (threadName)
-    , m_workers (workers)
+    : m_workers {workers}
+    , threadName_ {threadName}
+    , wakeCount_ {0}
+    , shouldExit_ {false}
 {
-    startThread ();
+    thread_ = std::thread {&Workers::Worker::run, this};
 }
 
 Workers::Worker::~Worker ()
 {
-    stopThread ();
+    {
+        std::lock_guard <std::mutex> lock {mutex_};
+        ++wakeCount_;
+        shouldExit_ = true;
+    }
+
+    wakeup_.notify_one();
+    thread_.join();
+}
+
+void Workers::Worker::notify ()
+{
+    {
+        std::lock_guard <std::mutex> lock {mutex_};
+        ++wakeCount_;
+    }
+    wakeup_.notify_one();
```

Comment:
Anthony Williams seems to prefer putting the notify outside the lock.  Here's a relevant blog post from him: https://www.justsoftwaresolutions.co.uk/threading/implementing-a-thread-safe-queue-using-condition-variables.html.  His justification is under the heading **Be careful when you notify**.

He has a similar justification on page 165 of **C++ Concurrency in Action**.